### ISSUE TITLE
AnalsysisCallback handler in Zinc is no longer thread-safe

### ISSUE THREAD

**Issue by dotta** [sbt/sbt#6198]
## steps

Compiling akka with Hydra can result in an exception originating in the sbt/zinc AnalysisCallback.

## problem

```
[info] Using 4 Hydra workers to compile Scala sources.
java.util.NoSuchElementException: key not found: akka.io.UdpConnected$ResumeReading$
  | => aat scala.collection.MapLike.default(MapLike.scala:235)
	at scala.collection.MapLike.default$(MapLike.scala:234)
	at scala.collection.AbstractMap.default(Map.scala:65)
	at scala.collection.MapLike.apply(MapLike.scala:144)
	at scala.collection.MapLike.apply$(MapLike.scala:143)
	at scala.collection.AbstractMap.apply(Map.scala:65)
	at sbt.internal.inc.AnalysisCallback.$anonfun$addProductsAndDeps$14(Incremental.scala:983)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:285)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:285)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:278)
	at scala.collection.mutable.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:50)
	at scala.collection.SetLike.map(SetLike.scala:105)
	at scala.collection.SetLike.map$(SetLike.scala:105)
	at scala.collection.mutable.AbstractSet.map(Set.scala:50)
	at sbt.internal.inc.AnalysisCallback.$anonfun$addProductsAndDeps$1(Incremental.scala:981)
	at scala.collection.TraversableOnce$folder$1$.apply(TraversableOnce.scala:187)
	at scala.collection.TraversableOnce$folder$1$.apply(TraversableOnce.scala:185)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:189)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:184)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:108)
	at sbt.internal.inc.AnalysisCallback.addProductsAndDeps(Incremental.scala:954)
	at sbt.internal.inc.AnalysisCallback.getAnalysis(Incremental.scala:889)
	at sbt.internal.inc.AnalysisCallback.dependencyPhaseCompleted(Incremental.scala:1035)
	at xsbt.Dependency$DependencyPhase.run(Dependency.scala:50)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1675)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1649)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1642)
	at scala.tools.nsc.Global$Run.compileFiles(Global.scala:1863)
	at com.triplequote.hydra.compiler.ScalaWorker$ScalaCompilationRound.compile(ScalaWorker.scala:123)
	at com.triplequote.hydra.compiler.ScalaWorker.compile(ScalaWorker.scala:72)
	at com.triplequote.hydra.WorkerThread.$anonfun$run$1(WorkerThread.scala:41)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at com.triplequote.hydra.compiler.utils.Utils$.timed(Utils.scala:34)
	at com.triplequote.hydra.WorkerThread.run(WorkerThread.scala:41)
```

## expectation

No exception.

## notes

The implementation of the AnalysisCallback has evolved and it seems it no longer thread-safe and it also now seem to expect sequential compilation. This wasn't the case in previous versions of sbt, as the AnalysisCallback implementation was made thread-safe in https://github.com/sbt/zinc/issues/405.

I'll look into how we can fix the problem. If you have suggestions, I'm all ears!

**Comment by eed3si9n** [comment]
We switched to using VirtualFiles everywhere in sbt 1.4 around https://github.com/sbt/zinc/pull/712, which could be related?

**Comment by dotta** [comment]
Hey Eugene! 👋 

Yes, I'm aware of the change to VirtualFile, and in fact we've been working on upgrading Hydra and the sbt-hydra plugin to support compilation with VirtualFile. I'm pretty sure the problem I've reported here is due to the changes made in the AnalysisCallback. Also, the exception is happening only from time to time, hinting at the fact that it's a race conditions. In fact, it's really the case that multiple Hydra workers are racing to compile the subset of sources assigned to each of them. As soon as one of them ends up calling the `AnalysisCallback.dependencyPhaseCompleted`, if the `classNames` are not fully populated (which I suspect can be the case as each Hydra worker proceeds at different speed), I believe we may end up hitting the problem reported in this ticket. This is at least the way I'm currently visualizing things, but I'm still digging so I might be wrong.

One question I have is how can we override the AnalysisCallback implementation in case the changes we need are too specific to Hydra? If I'm not mistaken, in sbt 1.3 and earlier the AnalysisCallback used to live in the compiler-bridge, while now it's in zinc-core. Hence, providing a different implementation seems a lot harder, if even possible.

**Comment by eed3si9n** [comment]
> If I'm not mistaken, in sbt 1.3 and earlier the AnalysisCallback used to live in the compiler-bridge, while now it's in zinc-core.

I don't think so. In general AnalysisCallback is a private communication between the injected compiler phases and Zinc core to build the Analysis file, and the implementation has been part of Zinc itself. For instance Zinc 1.3.5 has `private object AnalysisCallback`, and inside that we instantiate `new AnalysisCallback(...)`, which is `private final`.

https://github.com/sbt/zinc/blob/v1.3.5/internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala#L113-L137

**Comment by dotta** [comment]
You are absolutely right. I've good news: we managed to solve the origin of the problem and guarantee that `AnalysisCallback.dependencyPhaseCompleted` is called only once all Hydra workers have completed the sbt-dependency phase. All of this without touching the existing AnalysisCallback implementation.

However, I'd still like to restore thread-safety of the `AnalsysisCallback`. In particular, I was planning to use an `AtomicReference` for [`invalidationResults`](https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/Incremental.scala#L648-L650) and leverage the atomic operations provided by it.

If you are ok with the above, would you rather we close this ticket and create a new one that is solely focused on restoring thread-safety of the `AnalsysisCallback`? Or should we use this ticket for the change?

**Comment by eed3si9n** [comment]
> If you are ok with the above, would you rather we close this ticket and create a new one that is solely focused on restoring thread-safety of the `AnalsysisCallback`? 

Sounds good. I'm fine with using this issue for the thread-safety issue.

**Comment by eed3si9n** [comment]
Fixed in https://github.com/sbt/zinc/pull/957

**Comment by dotta** [comment]
This problem is fully fixed in the latest [sbt-hydra 2.3.0 release](https://twitter.com/triple_quote/status/1343255214137217024), in combination with using sbt 1.4.6 (as the latter includes the the sbt/zinc fix linked above by Eugene)
